### PR TITLE
TD-1220:Fixes to SelfAssessmentResults table usage

### PR DESCRIPTION
**Pull request by sarathlal-sarangadharan**
### JIRA link
https://hee-tis.atlassian.net/browse/TD-1220

### Description
1. Created history table **_SelfAssessmentResultSupervisorVerificationsHistory_** for  **_SelfAssessmentResultSupervisorVerifications_**
2. Deleted multiple records from tables **_SelfAssessmentResults_** & SelfAssessmentResultSupervisorVerifications
3. Created a Unique constraint to the **_SelfAssessmentResults_** table on DelegateUserID, CompetencyID, AssessmentQuestionID
4. Altered queries used logic to get latest record from  SelfAssessmentResults table.

### Screenshots
**_Before code changes:_**
image
image
image
image

**_After code changes:_**
SelfAssessmentResultSupervisorVerifications table changes:
image

**SelfAssessmentResults table changes**
image

**Deleted duplicate records**
image

**submit changes:**
image

**select data count :**
image

**SelfAssessmentReportDartaService**
image

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors (see [info on Text Editor settings](https://hee-tis.atlassian.net/wiki/spaces/TP/pages/3546185813/DLS+Dev+Process) to avoid whitespace changes)
- [ ] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript
- [ ] Tested any Views or partials created or changed with [Wave Chrome plugin](https://chrome.google.com/webstore/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh/related) and addressed any valid accessibility issues
- [ ] Updated/added documentation in [Confluence](https://hee-tis.atlassian.net/wiki/spaces/TP/pages/3546939432/DLS+Code) and/or [GitHub Readme](https://github.com/TechnologyEnhancedLearning/DLSV2/blob/master/README.md). List of documentation links added/changed:
  - [doc_1_here](link_1_here)
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken
- [ ] Scanned over my pull request in GitHub and addressed any warnings from the GitHub Build and Test checks.

### CODE REVIEW

**Comment by sarathlal-sarangadharan** [comment]
> @sarathlal-sarangadharan, I was expecting the system versioning to be switched on before we delete the SelfAssessmentResultSupervisorVerifications records so that we don't lose the record of them. Will this work?

Hi @kevwhitt-hee  , I tried this already. But was getting below error when we use system versioning before data delete.
image

Will check this again and update you.

**Comment by kevwhitt-hee** [comment]
> > @sarathlal-sarangadharan, I was expecting the system versioning to be switched on before we delete the SelfAssessmentResultSupervisorVerifications records so that we don't lose the record of them. Will this work?
> 
> Hi @kevwhitt-hee , I tried this already. But was getting below error when we use system versioning before data delete. image
> 
> Will check this again and update you.

Looks like we need to add begin / end transactions to the steps - https://dba.stackexchange.com/questions/211467/error-updating-temporal-tables?

**Comment by sarathlal-sarangadharan** [comment]
> > > @sarathlal-sarangadharan, I was expecting the system versioning to be switched on before we delete the SelfAssessmentResultSupervisorVerifications records so that we don't lose the record of them. Will this work?
> > 
> > 
> > Hi @kevwhitt-hee , I tried this already. But was getting below error when we use system versioning before data delete. image
> > Will check this again and update you.
> 
> Looks like we need to add begin / end transactions to the steps - https://dba.stackexchange.com/questions/211467/error-updating-temporal-tables?

Hi @kevwhitt-hee ,Getting error after using Begin/end transaction. Seems like we need to build exception handling to retry till the transaction is successful. Working on it.

**Comment by sarathlal-sarangadharan** [comment]
> > > > @sarathlal-sarangadharan, I was expecting the system versioning to be switched on before we delete the SelfAssessmentResultSupervisorVerifications records so that we don't lose the record of them. Will this work?
> > > 
> > > 
> > > Hi @kevwhitt-hee , I tried this already. But was getting below error when we use system versioning before data delete. image
> > > Will check this again and update you.
> > 
> > 
> > Looks like we need to add begin / end transactions to the steps - https://dba.stackexchange.com/questions/211467/error-updating-temporal-tables?
> 
> Hi @kevwhitt-hee ,Getting error after using Begin/end transaction. Seems like we need to build exception handling to retry till the transaction is successful. Working on it.

Hi @kevwhitt-hee , I separated the logics to different classes now. Please review and share your thought. It is working as expected and deleted records are stored in history table.
image

**Comment by kevwhitt-hee** [comment]
Approved but I'm going to check out the branch and test before merging. Thanks @sarathlal-sarangadharan

**Comment by kevwhitt-hee** [comment]
This is looking good, @sarathlal-sarangadharan. There are a few more places where we are retrieving the latest SelfAssessmentResults record where we no longer need to be:

- SupervisorService.cs 
  - GetSelfAssessmentResultSummary - lines 902, 916, 932, 964, 980
  - GetCandidateAssessmentSupervisorVerificationSummaries - line 1031

- CandidateAssessmentExportDataService.cs
 - GetCandidateAssessmentExportSummary - lines 24, 38, 54, 70, 86, 102

Please could these be updated too?

**Comment by sarathlal-sarangadharan** [comment]
> This is looking good, @sarathlal-sarangadharan. There are a few more places where we are retrieving the latest SelfAssessmentResults record where we no longer need to be:
> 
> * SupervisorService.cs
>   
>   * GetSelfAssessmentResultSummary - lines 902, 916, 932, 964, 980
>   * GetCandidateAssessmentSupervisorVerificationSummaries - line 1031
> * CandidateAssessmentExportDataService.cs
> * GetCandidateAssessmentExportSummary - lines 24, 38, 54, 70, 86, 102
> 
> Please could these be updated too?

Hi @kevwhitt-hee ,Sure I will work on this.

File: DigitalLearningSolutions.Data/DataServices/SelfAssessmentDataService/CompetencyDataService.cs [hunk 1]
```diff
@@ -423,15 +406,9 @@ UPDATE SelfAssessmentResults
...
-                            INSERT INTO SelfAssessmentResults
-                                ([SelfAssessmentID]
-                                ,[CompetencyID]
-                                ,[AssessmentQuestionID]
-                                ,[Result]
-                                ,[DateTime]
-                                ,[SupportingComments]
-                                ,[DelegateUserID])
-                            VALUES(@selfAssessmentId, @competencyId, @assessmentQuestionId, @result, GETUTCDATE(), @supportingComments,@delegateUserId)
+                            UPDATE SelfAssessmentResults
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Data/DataServices/SelfAssessmentDataService/CompetencyDataService.cs:409]: If we update a `SelfAssessmentResults` record we need to delete any related `SelfAssessmentResultSupervisorVerifications` record.

**Review by kevwhitt-hee — CHANGES_REQUESTED** [review]
Noticed a scenario that we need to cope with - 

- Learner submits a self assessment result
- Supervisor confirms result (SelfAssessmentResultSupervisorVerifications record)
- Learner submits a new self assessment result for same and we update it

We need to delete the SelfAssessmentResultSupervisorVerifications associated with the original result.

**Comment by sarathlal-sarangadharan** [comment]
> > This is looking good, @sarathlal-sarangadharan. There are a few more places where we are retrieving the latest SelfAssessmentResults record where we no longer need to be:
> > 
> > * SupervisorService.cs
> >   
> >   * GetSelfAssessmentResultSummary - lines 902, 916, 932, 964, 980
> >   * GetCandidateAssessmentSupervisorVerificationSummaries - line 1031
> > * CandidateAssessmentExportDataService.cs
> > * GetCandidateAssessmentExportSummary - lines 24, 38, 54, 70, 86, 102
> > 
> > Please could these be updated too?
> 
> Hi @kevwhitt-hee ,Sure I will work on this.

Hi @kevwhitt-hee ,
1. GetSelfAssessmentResultSummary - lines 902, 916, 932, 964, 980
Before code change:
image

After code change::
image

2. GetCandidateAssessmentSupervisorVerificationSummaries - line 1031
Before code change:

After code change:
image


3. CandidateAssessmentExportDataService.cs
      GetCandidateAssessmentExportSummary - lines 24, 38, 54, 70, 86, 102

Before code change:
image

After code change:
image

**Comment by sarathlal-sarangadharan** [comment]
> * Learner submits a self assessment result
> * Supervisor confirms result (SelfAssessmentResultSupervisorVerifications record)
> * Learner submits a new self assessment result for same and we update i

@kevwhitt-hee  Working on this scenario. Will connect you for any doubt.

File: DigitalLearningSolutions.Data/DataServices/SelfAssessmentDataService/CandidateAssessmentExportDataService.cs [hunk 2]
```diff
@@ -82,10 +70,7 @@ public CandidateAssessmentExportSummary GetCandidateAssessmentExportSummary(int
                               sar1.AssessmentQuestionID = caqrr1.AssessmentQuestionID RIGHT OUTER JOIN
                               SelfAssessmentStructure AS sas1 INNER JOIN
                               CandidateAssessments AS ca1 ON sas1.SelfAssessmentID = ca1.SelfAssessmentID INNER JOIN
-                              CompetencyAssessmentQuestions AS caq1 ON sas1.CompetencyID = caq1.CompetencyID ON sar1.ID =
-                                  (SELECT MAX(ID) AS Expr1
-                                  FROM    SelfAssessmentResults AS sar2
-                                  WHERE (CompetencyID = caq1.CompetencyID) AND (AssessmentQuestionID = caq1.AssessmentQuestionID) AND (DelegateUserID = ca1.DelegateUserID) AND (SelfAssessmentID = ca1.SelfAssessmentID)) LEFT OUTER JOIN
+                              CompetencyAssessmentQuestions AS caq1 ON sas1.CompetencyID = caq1.CompetencyID ON sar1.SelfAssessmentID =sas1.SelfAssessmentID and sar1.CompetencyID=sas1.CompetencyID LEFT OUTER JOIN
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Data/DataServices/SelfAssessmentDataService/CandidateAssessmentExportDataService.cs:73]: and again

File: DigitalLearningSolutions.Data/DataServices/SelfAssessmentDataService/CandidateAssessmentExportDataService.cs [hunk 3]
```diff
@@ -98,10 +83,7 @@ public CandidateAssessmentExportSummary GetCandidateAssessmentExportSummary(int
...
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Data/DataServices/SelfAssessmentDataService/CandidateAssessmentExportDataService.cs:86]: and again

**Review by kevwhitt-hee — CHANGES_REQUESTED** [review]
I don't think we are limiting the join to SelfAssessmentResults enough in this query. See code comments.

**Comment by sarathlal-sarangadharan** [comment]
> I don't think we are limiting the join to SelfAssessmentResults enough in this query. See code comments.

Hi @kevwhitt-hee , Amended the query . Please review and also please find test evidences. Now count matches with old query.

Original query result:
image

Latest query result:
image

SupervisorService.cs
Original query result:
image

Latest query result:
image

**Pull request closed without merging**